# Algoritmo 4: (1+1)-EE
TODO: Entrada
- Valor inicial de o (desviación estándar), constante c, número máximo de generaciones G y parámetros de entrada para el problema que se quiere resolver.

TODO: Salida
- Mejor solución encontrada.

TODO: Pasos del algoritmo (1+1)-EE

**Primero:** crea un individuo inicial x que es conocido como **padre**
**Después:** genera un hijo y = x + z, donde z es un vector generado con números aleatorios utilizando una distribución normal con media 0 y desviación estándar $\sigma:$ $z_{i}$ = N(0, $\sigma$). El valor de $\sigma$ controla qué tanto son perturbadas las variables de decisión a la hora de generar un hijo. Este valor se ajusta a cada determinado número de iteraciones, es decir se auto-adapta, utilizando **la regla del éxito 1/5**. Esta regla enuncia que la tasa de mutaciones exitosas (es decir, mutaciones en las que el hijo es mejor que el padre), debería ser exactamente de 1/5. La regla del éxito del 1/5 es aplicada cada k iteraciones de la siguiente forma:

$\sigma$ = {$\sigma$/c Sí p_{s} > 1/5, $\sigma$*c Sí p_{s} < 1/5, o si p_{s} = 1/5}

Dónde $p_{s}$ es el porcentaje de mutaciones exitosas y 0.817 <= c <= 1 Si se tiene demasiadas mutaciones
exitosas ($p_{s}$ > 1/5), o debería ser incrementado para realizar una búsqueda más amplia en el espacio de soluciones.

Si se tienen muy pocas mutaciones exitosas ($p_{s}$ < 1/5), o debería ser decrementado para para concentrar la búsqueda alrededor del individuo actual. Finalmente, padre e hijo compiten y el que tenga una mejor aptitud será el padre de la siguiente iteración.

# REPRESENTACIÓN

Cada individuo almacena tanto las variables del problema que se desea resolver como los parámetros que utiliza la EE en el operador de mutación. Por ejemplo: $[x_{1}, ..., x_{n}, \sigma_{1}, ..., \sigma_{m}]$
$x_{1}, ..., x_{n}$ son valores flotantes asociados a cada una de las n variables de decisión del problema.
$\sigma_{1}, ..., \sigma_{m}$ definen las formas de las distribuciónes utilizadas para mutar a los individuos (tamaños de paso) **regularmente se utiliza m=1 ó m=n;**

# SELECCIÓN DE PADRES
Las EEs seleccionan a los individuos que serán padres de manera aleatoria, utilizando una distribución uniforme.
los individuos tiene la misma probabilidad de ser padres, sin importar su aptitud.

# CRUZA
Regularmente, el operador de recombinación o cruza utiliza dos individuos (padres) para generar un nuevo individuo (hijo).Tenemos: 
    - Recombinación discreta 
    - Recombinación intermedia:

**DISCRETA:** Recorre uno a uno los componentes de cada padre y se decide de manera aleatoria si la información del nuevo individuo se toma del padre 1 ó del padre 2.

**INTERMEDIA:** La recombinación intermedia genera al nuevo individuo utilizando el promedio de ambos padres para cada componente. Supongamos que a partir de los padres X e Y, se va a generar al hijo Z.
Quedando los operadores de la siguiente forma:
- Recombinación discreta: $Z_{i} = X_{i}$ ó $Y_{i}$ (elegido aleatoriamente)
- Recombinación intermedia: $Z_{i} = X_{i} + Y_{i} / 2$

# MUTACIÓN
El operador de mutación utiliza una distribución normal. 
Este tipo de distribución requiere de dos parámetros: la media u y la desviación estándar $\sigma$. 
Regularmente, u es igual a cero (0), o se va estableciendo conforme la búsqueda avanza. 

**Las variables de decisión se mutan de la siguiente forma:** 

$$
X^{´}_{i} = X_{i} + \sigma^{´}_{i} * N(0,1)$$
Donde N(0,1) genera un número aleatorio utilizando una distribución normal con 
media 0 y desviación estándar 1. 

**MUTACIÓN CON UN ÚNICO TAMAÑO DE PASO**

Cuando se tiene un único valor o en cada individuo: $[x_{1}, ..., x_{n}, 0]$, el valor o se muta de la siguiente manera: 

$$
\sigma^{'} = \sigma * e^{t*N(0,1)}
$$

Dónde N(0, 1) genera un número aleatorio utilizando una distribución 
normal con media 0 y desviación estándar 1. t es un parámetro de entrada al algoritmo
que es visto como una tasa de aprendizaje. Regularmente se usa. t = 1/$sqrt(n)$, dónde n es el número de variables del problema.

**Para evitar valores muy pequeños en $\sigma$, si $\sigma^{'}$ < $E_{0}$ se hace $\sigma^{'}$ = $E_{0}$ 

**SELECCIÓN DE SOBREVIVIENTES**
Las EEs suelen utilizar algunos de los siguinetes esquemas de selección de sobrevivientes: 

- **SELECCIÓN (u, $\lambda$)**. A partir de u posibles padres, se generan $\lambda$ hijos, con $\lambda$ >= u. De los $\lambda$ hijos, se seleccionan a los u mejores.Estos individuos formarán parte de la siguiente iteración. Este tipo de selección se recomienda en problemas en los que el óptimo es dinámico o cuando la función es multmimodal. Se recomienda utilizar $\lambda$ = 7u

- **SELECCIÓN (u+$\lambda$)**. A partir de u posibles padres, se generan $\lambda$ hijos. Se unen las poblaciones de padres e hijos y se seleccionan a los mejores u individuos. Estos individuos formarán parte de las siguiente iteración. Este tipo se considera elitista. 


**REPRESENTACIÓN DE UN INDIVIDUO** 
Cada individuo será visto como un arreglo de tamaño n + 1. Las primeras
n componentes corresponden a las variables de decisión del problema $x_{i}$ y
la última componente corresponde con el tamaño de paso. En este caso
usaremos un único tamaño de paso para todas las variables de decisión.
Si n = 2, un posible individuo es: $[ 10.3, 7.84, 0.84]$

**POBLACIÓN INICIAL** 
Generamos un individuo de la siguiente forma: Utilizando una distribución
uniforme, generamos n números aleatorios en el intervalo $[ 30, 30]$ y los guardamos en las primeras
n posiciones de un arreglo de tamaño n + 1. Posteriormente, generamos otro número aleatorio en el intervalo (0, 1) y lo guardamos en la última posición de nuestro arreglo. Repetimos el mismo procedimiento
para generar la cantidad de individuos que se deseen. En este caso usaremos
una población de tamaño μ = 100.


# Ejercicios
[X] Implementar el algoritmo 4 para resolver la función de Ackley


---


$$
min f(\overrightarrow{x}) = -20 \ast exp \left( -0.2 \ast \sqrt{\frac{1}{n} \sum_{i=1}^{n}x^{2}_{i}}\right) - exp \left( \frac{1}{n} \sum_{i=1}^{n} cos(2 \pi x_{i})  \right) + 20 + e
$$

> La función $f$ es conocida como la **función de Ackley**, su dominio es $|x_{i}| \leq 30$ y su mínimo global está en $x_{i} = 0$ y $f(\overrightarrow{x}) = 0.0$


[X] Implementar un algoritmo basado en EEs para resolver la función de Ackley. El objetivo es mejorar los resultados obtenidos con el diseño propuesto en el ejemplo

[X] Permitir al usuario realizar M ejecuciones del algoritmo de EE, implementado en el punto anterior, para resolver la instancia del problema de Ackley. Reportar: Mejor solución, Peor solución, La media de soluciones, desviación estandar del valor de la función objetivo

[x] Realizar una tabla comparativa que reporte los resultados de
las tres versiones de EEs implementadas: (1 + 1)-EE, (μ,$\lambda$)-EE y la versión propuesta. Se deben reportar los resultados para problemas con 2, 5, 7 y 10 variables.

# Ejercicio 1 
---
Implementar en el lenguaje de su preferencia, el algoritmo 4 para resolver la función de Ackley

In [4]:
from math import exp
from math import sqrt
from math import cos
from math import e
from math import pi
import numpy as np
from numpy import std
from random import randint

In [149]:
# TODO: Función de Ackley
def f(x, n): 
  """
  x = list
  n = int, num of decisition variables
  return: float number
  """
  #print("f", x)
  #print("FF", x[0])
  # Para efectuar la primer sumatoria
  sum1 = 0
  for i in range(n): 
    sum1 += x[i]**2
  sumatoria1 = (1/n)*sum1
  # Para efectuar la segunda sumatoria
  sum2 = 0
  for i in range(n): 
    sum2 += cos(2*pi*x[i])
  sumatoria2 = (1/n)*sum2

  return (-20 * exp(-0.2*sqrt(sumatoria1)) - exp(sumatoria2) + 20 + e)

In [150]:
def fitness(x, n):
  """
  x = list
  n = int, num of decision variables
  return: float number
  """
  return -f(x, n)

# Representación del individuo

Un arreglo de valores flotantes de tamaño 4. Las primeras dos componentes corresponden a las dos variables del problema. Las siguentes dos componentes corresponden a los tamaños de paso para la mutación de cada una de las variables del problema.

x1 = abs(np.random.uniform(-30, 30))

x2 = abs(np.random.uniform(-30, 30))

$\sigma$1 = np.random.uniform(0,1)

$\sigma$2 = np.random.uniform(0,1)


X = [x1, x2, $\sigma$1, $\sigma$2]

# Población Inicial

In [151]:
# En general, el arreglo de representación 
# del individuo es de tamaño 2n
# Donde
# Las primer mitad de componentes corresponden a las n variables del problema.
# Las otra mitad de componentes corresponden a los n tamaños de paso para la mutación
# de cada una de las variables del problema.
def getInitialPopulation(n, mu=False):
    """
    mu = int, poblation size
    n = int, num of decision variables
    return: list with initial population and their fitness by guy
    """
    if mu != False:
      population = []
      for i in range(mu):
        #Generamos un individuo
        guy = np.concatenate((abs(np.random.uniform(-30, 30, n)), np.random.uniform(0, 1, 1)), axis=0)

        #Calculamos la aptitud del individuo
        guy = [guy, fitness(guy[:n], n)]
        #print(guy)
        #break

        #Agregamos al individuo a la población "parents"
        population.append(guy)
      
      # Regresamos la población generada
      # Arreglo de arreglos, donde un indice corresponde a un individuo formado como: 
      # mitad variables del problema, mitad variables de mutación y fitness
      return population
    
    else:
      left = np.random.uniform(-30,30,n)
      right = np.random.uniform(0,1,1)
      left_fitness = f(left, n) 
      guy =  np.concatenate((left, right), axis=0)
      guy = [guy, left_fitness]
      
    return guy

    
    

# Mutación 

In [153]:
def mutation(guy, n):
    """
      guy = list, guy number any
      n = int, num of decision variables
      return: list, new guy
    """

    # Mutamos los componentes de nuestro arreglo
    # TODO: Recordemos que o (sigma) se va estableciendo conforme la búsqueda avanza
    # por lo que le asignamos un primer valor
    o = guy[0][-1]
    for i in range(n):
      guy[0][i] = guy[0][i]+o*np.random.normal(0,1)
      

    # Revisamos que las variables de decisión estén dentro de los límites
    # De lo contrario las igualamos al limite
    #guy[:n < -30] = -30
    #guy[:n > 30] = 30
    #print(guy)
    for i in range(n):
      if guy[0][i] < -30: guy[0][i] = -30
      elif guy[0][i] > 30: guy[0][i] = 30
    #print(guy)

    guy_fitness = fitness(guy[0][:n], n)
    # Creamos el nuevo Individuo y lo devolvemos
    return [[guy, guy_fitness]][0][0]

# Main: (1+1)-EE

In [155]:
def one_one_EE(o, c, G, k, n):
    """
    o: float - init value sd 
    c: int - constant value
    G: int - max number of generations
    k: int - Actualizacion generaciones
    n: int, num of decision variables
    """
    t = 0
    m_successful = 0
    # Obtener solución inicial x
    x = getInitialPopulation(n)
    # Evaluar f(x)
    #x_Eval = fitness(x, n) # TODO: Revisando el pseudocódigo de (1+1)-EE, no me queda claro la finalidad de sta evaluación
    while t <= G: 
      # Mutar solución x_head = x´i = xi + o *Ni(0,1)
      x_head = mutation(x, n)
      
      if x_head  < x: 
        x = x_head
        m_successful += 1
      
      t += 1
      if t % k == 0: 
        Ps = m_successful/k
        
        if Ps > 1/5:
          o = o/c

        elif Ps < 1/5: 
          o = o*c

        m_successful = 0

    return x_head
        

In [156]:
#!pip install python-dotenv

In [165]:
from dotenv import dotenv_values
if __name__=="__main__":
  variables = dotenv_values("11-EE.txt")
  O = float(variables["11eeO"])
  C = float(variables["11eeC"])
  GG = float(variables["11eeG"])
  K = float(variables["11eeK"])
  N = int(variables["11eeN"])
  print(N)
  one_one_EE(o=O, c=C, G=GG, k=K, n=N)

  M = int(input("Número de ejecuciones: "))

  solutions = {}
  conteo = 0
  while conteo != M:
    current = one_one_EE(o=O, c=C, G=GG, k=K, n=N)
    xs = current[0][:-1]
    fx = current[1]
    solutions[tuple(xs)] = fx
    conteo += 1  

  
  # TODO: Ordenamos de menor a mayor por fitness
  solutions = sorted(solutions.items(), key=lambda x: x[1])
  #print(solutions)
  #print(solutions[0]) #best
  #print(solutions[-1]) #faild
  
  # MEJOR SOLUCION
  print(f"\nMEJOR SOLUCION\nXs: {solutions[0][0][:-1]}\nfx: {solutions[0][1]}")
  # PEOR SOLUCION 
  print(f"\nPEOR SOLUCION\nXs: {solutions[-1][0][:-1]}\nfx: {solutions[-1][1]}")
  # MEDIANAS
  print("\nMEDIANA: Xs&fx")
  tmp_X = []
  tmp_F = []
  for i in range(len(solutions)): 
    tmpX = solutions[i][0][:-1]
    tmpF = solutions[i][1]
    tmp_X.append(tmpX)
    tmp_F.append(tmpF)


  num_ele = len(tmp_X[0])
  aux = [0 for i in range(num_ele)]
  
  for element in tmp_X:
    for i in range(num_ele):
      aux[i] += element[i]

  for i in range(len(aux)):
    aux[i] = aux[i]/len(tmp_X)

  # Media de ambos
  print(f"Xs&fx: {aux} & {sum(tmp_F)/len(tmp_F)}")
  
  # Media del valor de la f objetivo
  print("\nMEDIA FUNCION OBJETIVO")
  print(f"Media: {sum(tmp_F)/len(tmp_F)}")

  # Desviación estánda de la f objetivo
  print("\nDESVIACION ESTANDAR FUNCION OBJETIVO")
  print(f"SD: {np.std(tmp_F)}")
    

10
Número de ejecuciones: 5000

MEJOR SOLUCION
Xs: (-2.044246570159379, 17.88231351065758, -6.052851320747088, -1.1693274828589078, -0.9154921722675008, 7.962487770169621, -2.9105446028522426, -6.105693861656952, 1.1571605460194185)
fx: 17.547398709030844

PEOR SOLUCION
Xs: (15.130841356357774, 21.870701365438478, 29.77423858024341, 28.481443809529747, 26.140565658607038, -28.105194874326962, -25.515260304844062, -2.2332799251082722, -30.0)
fx: 21.94152537433068

MEDIANA: Xs&fx
Xs&fx: [0.010372363820764036, -0.02250920550874845, -0.18706328010297152, 0.4985264746541095, 0.08681879516998701, -0.27355355064089315, 0.08632618368508485, -0.36110568281307387, 0.031095853796096355] & 20.946555311456187

MEDIA FUNCION OBJETIVO
Media: 20.946555311456187

DESVIACION ESTANDAR FUNCION OBJETIVO
SD: 0.4664691168459052


# Implementación: ($\mu$, $\lambda$)-EE
---

Dado que para realizar la tabla comparativa necesito este algoritmo, he de implementarlo aquí con apoyo del manual

In [169]:
import numpy as np
from copy import copy
from math import e, exp, sqrt, pi, cos
from random import choices

## Representación del individuo

In [171]:
# TODO: Representación de un individuo 
# Representación: Cada individuo será visto como un arreglo de valroes flotantes de tamaño n+1
# Las primeras n componentes corresponden a las variables de decisión del problema 
# y las siguiente componente corresponde al tamaño de paso para la mutación de las variables del problema


## Aptitud

In [172]:
def f(x): 
  part1 = 20*exp(-0.2*sqrt(np.average(x**2)))
  part2 = exp(np.average(np.array([cos(i) for i in 2*pi*x])))
  return -part1-part2+20+e

def fitness(x):
  return -f(x)

## Población inicial 
---
Por cada individuo requerido, se utiliza una distribución uniforme para generar números aleatorios. Primero se generan n números en el intervalo [30, 30], cada uno corresponde a una variable de decisión. Posteriormente, un número aleatorio en el intervalo (0, 1) que
corresponde al tamaño de paso.

In [174]:
def getInitialPopulation(mu, n):
  population = []
  for i in range(mu):
    #Generamos un individuo
    p = np.concatenate((np.random.uniform(-30, 30, n),
    np.random.uniform(0, 1, 1)))
    #Calculamos la aptitud del individuo
    p = [p, fitness(p[:n])]
    #Agregamos al individuo a la población "parents"
    population.append(p)
    
  return population

## Cruza

Recombinación discreta local. Utilizando una distribución uniforme, generamos un número aleatorio en el intervalo (0, 1) por cada componente de nuestro individuo. Recordemos que las primeras n componentes del individuo corresponden a sus variables de decisión y la última componente a su tamaño de paso. Si el valor es menor a 0.5, el hijo
se queda con el valor del primer padre. De lo contrario, el hijo se queda con el valor del
segundo padre.

In [176]:
def localDiscreteCrossover(parent_1, parent_2):
  N = len(parent_1)
  #Creamos un vector con N componentes en el intervalo
  # (0,1) con distribución uniforme
  z = np.random.uniform(0, 1, N)
  child = np.array([0.0]*N)
  #Asignamos el valor del padre correspondiente
  for i in range(N):
    if z[i] < 0.5:
      child[i] = parent_1[i]
  else:
    child[i] = parent_2[i]
  
  #Regresamos el individuo generado
  return child

## Mutación 
para mutar necesitamos: 

- tamaño de paso: $\sigma^{´} = \sigma * e^{t*N(0,1)}$

Si el valor de $\sigma^{´}$ es de menor a $\epsilon_{0}$ hacemos que 
$\sigma^{'} = \epsilon_{0}$ Posteriormente utilizamos $\sigma^{'}$ para mutar las variables de decisión


Si n = 2, nuestras variables de deción mutadas quedan de la siguiente forma: 

$$
x^{'}_{1} = x_{1} + \sigma^{'} * N(0,1)
$$

$$
x^{'}_{2} = x_{2} + \sigma^{'} * N(0,1)
$$

Revisamos que las variables de decisión generadas estén dentro del rango que indica el problema. En caso de que no, se les asigna el valor del límite que rebasó. Construimos el individuo: 

$$
i^{'} = [x^{'}_{1}, x^{'}_{2}, \sigma^{'}]
$$

In [178]:
def mutation(i, n, epsilon, tau):
  #Mutamos el tamaño de paso
  mutation_sigma = i[-1]*(e**(tau*np.random.normal(0, 1, 1)))
  #Verificamos que el nuevo valor no sea menor a épsilon
  if mutation_sigma[0] < epsilon:
    mutation_sigma[0] = epsilon

  #Mutamos las variables de decisión
  mutation_x = i[:n] + (mutation_sigma[0]*np.random.normal(0, 1, n))

  #Revisamos que estén dentro de los límites
  mutation_x[mutation_x < -30] = -30
  mutation_x[mutation_x > 30] = 30

  return np.concatenate((mutation_x, mutation_sigma))

## Crear hijo usando cruza y mutación 

In [180]:
def createChild(parent_1, parent_2, n, epsilon, tau):
  new_ind = localDiscreteCrossover(parent_1, parent_2)
  new_ind = mutation(new_ind, n, epsilon, tau)

  #Regresamos el nuevo individuo, junto con su valor de apitud
  return [new_ind, fitness(new_ind[:n])]

## Main: $(\mu, \lambda)$ - EE
---
Se usa una selección ($μ, \lambda$) donde μ=100 y $\lambda$=700. Se elegirán dos padres al azar para
generar un hijo nuevo y este proceso se llevará a cabo 700 veces. De los hijos resultantes,
se elegirán a los 100 mejores que pasarán a la siguiente generación.

In [182]:
def EvolutionStrategies(n, G, epsilon=0.01, tau=None, mu_=100, lambda_=700):
  #Si el usuario no define tau,
  #se usa el valor recomendado en la literatura
  if tau == None:
    tau = 1/sqrt(n)

  population = getInitialPopulation(mu_, n)
  #Generamos los índices de los individuos para poder elegirlos
  #aleatoriamente
  population_idx = range(mu_)

  #Repetimos el proceso por G generaciones
  for _ in range(G):
    offspring = []
    #Generamos lambda_ hijos
    for i in range(lambda_):
      #Seleccionamos a los padres
      parents_idx = choices(population_idx, k=2)
      parent_1 = population[population_idx[0]][0]
      parent_2 = population[population_idx[1]][0]
      #Creamos el hijo y lo añadimos a la población de hijos
      offspring.append( createChild(parent_1, parent_2, n, epsilon, tau) )
    
    #Ordenamos del mejor individuo al peor individuo
    offspring.sort(key=lambda x: x[-1], reverse=True)
    #Seleccionamos a los mu_ mejores
    population = offspring[:mu_].copy()

  #Regresamos el mejor individuo
  return population[0][0], -population[0][1]

In [191]:
from dotenv import dotenv_values
if __name__=="__main__":
  variables = dotenv_values("11-EE.txt")
  N = int(variables["mulambdaN"])
  MU_ = int(variables["mulambdaMu"]) #tamaño de la población
  LAMBDA_ = int(variables["mulambdaLambda"]) #Cantidad de hijos que se generan por generación
  GG = int(variables["mulambdaG"]) #número de generaciones
  TAU = 1/sqrt(n) #tasa de aprendizaje
  EPSILON = float(variables["mulambdaEpsilon"]) #valor mínimo de sigma (tamaño de paso)

  print(N)  
  M = int(input("Número de ejecuciones: "))

  solutions = {}
  conteo = 0
  while conteo != M:
    current_xs, current_fx = EvolutionStrategies(n=N, G=GG, epsilon=EPSILON,tau=TAU, mu_=MU_, lambda_=LAMBDA_)
    xs = current_xs[:N]
    fx = current_fx
    solutions[tuple(xs)] = fx
    conteo += 1  

  
  # TODO: Ordenamos de menor a mayor por fitness
  solutions = sorted(solutions.items(), key=lambda x: x[1])
  #print(solutions)
  #print(solutions[0]) #best
  #print(solutions[-1]) #faild
  
  # MEJOR SOLUCION
  print(f"\nMEJOR SOLUCION\nXs: {solutions[0][0][:-1]}\nfx: {solutions[0][1]}")
  # PEOR SOLUCION 
  print(f"\nPEOR SOLUCION\nXs: {solutions[-1][0][:-1]}\nfx: {solutions[-1][1]}")
  # MEDIANAS
  print("\nMEDIANA: Xs&fx")
  tmp_X = []
  tmp_F = []
  for i in range(len(solutions)): 
    tmpX = solutions[i][0][:-1]
    tmpF = solutions[i][1]
    tmp_X.append(tmpX)
    tmp_F.append(tmpF)


  num_ele = len(tmp_X[0])
  aux = [0 for i in range(num_ele)]
  
  for element in tmp_X:
    for i in range(num_ele):
      aux[i] += element[i]

  for i in range(len(aux)):
    aux[i] = aux[i]/len(tmp_X)

  # Media de ambos
  print(f"Xs&fx: {aux} & {sum(tmp_F)/len(tmp_F)}")
  
  # Media del valor de la f objetivo
  print("\nMEDIA FUNCION OBJETIVO")
  print(f"Media: {sum(tmp_F)/len(tmp_F)}")

  # Desviación estánda de la f objetivo
  print("\nDESVIACION ESTANDAR FUNCION OBJETIVO")
  print(f"SD: {np.std(tmp_F)}")  
  

10
Número de ejecuciones: 10

MEJOR SOLUCION
Xs: (-0.007856224531891857, -0.004900643004742735, 0.0023537358139212325, -0.0008414020724393399, 0.002150847399051097, 0.0046581898315798944, -0.0012226454308359954, 0.001062616874116001, 0.002619402286077113)
fx: 0.014992679304719037

PEOR SOLUCION
Xs: (-0.0047211243877690635, -0.005492142626463613, -0.002548911092052195, -0.0015819740857767558, 0.0019770234868002746, 0.0067102071411357725, 0.005643268025221114, 0.007040333883980928, -0.002103292223683865)
fx: 0.020937350486651685

MEDIANA: Xs&fx
Xs&fx: [-0.0016897703135800156, -0.0015113848830057494, -0.001775757451882999, 0.0014822654723971043, 0.002669917249006662, 0.001753149609581519, 0.00021638853540399913, -0.0006085943719727511, 0.0001282185673733577] & 0.018763338761734704

MEDIA FUNCION OBJETIVO
Media: 0.018763338761734704

DESVIACION ESTANDAR FUNCION OBJETIVO
SD: 0.0021167310260617705


# Ejercicio 2
---
Implementar, en el lenguaje de su preferencia, un algoritmo basado en EEs para resolver la función de Ackley. El objetivo es mejorar los resultados obtenidos con el diseño propuesto 


---
Para resolver esta parte del ejercicio voy a hacer uso de la implementación $(\mu, \lambda)-EE$ sin embargo voy a modificar dos partes, la __cruza__ y la __mutación__ ya que considero que son las partes fundamentales de un algoritmo basado en EEs y por lo tanto no he de crearlo desde cero sino que puedo reutilizar algunas funciones que ya hay en este documento

In [192]:
import numpy as np
from copy import copy
from math import e, exp, sqrt, pi, cos
from random import choices

## Función y calculo de aptitud

In [193]:
def f(x): 
  part1 = 20*exp(-0.2*sqrt(np.average(x**2)))
  part2 = exp(np.average(np.array([cos(i) for i in 2*pi*x])))
  return -part1-part2+20+e

def fitness(x):
  return -f(x)

## Población inicial 

In [194]:
def getInitialPopulation_plus(mu, n):
  population = []
  for i in range(mu):
    #Generamos un individuo
    p = np.concatenate((np.random.uniform(-30, 30, n),
    np.random.uniform(0, 1, 1)))
    #Calculamos la aptitud del individuo
    p = [p, fitness(p[:n])]
    #Agregamos al individuo a la población "parents"
    population.append(p)
    
  return population

## Cruza modificada

In [195]:
def localDiscreteCrossover_modified(parent_1, parent_2):
  N = len(parent_1)
  #Creamos un vector con N componentes en el intervalo
  # generamos a los padres a partir de una distribución de chi cuadrada
  # de tal forma que privilegiamos más a unos números que a otros
  z = np.random.chisquare(1, N)
  
  child = np.array([0.0]*N)
  
  #Asignamos el valor del padre correspondiente
  # En este caso, daremos más libertad de elección, es decir
  # la ampliaremos a 0.8 a su vez, el hijo no sé quedara con el 
  # valor del primer padre sino que se quedara con un promedio 
  # de ambos, que básicamente sería recombinación intermedia
  for i in range(N):
    if z[i] < 0.8:
      child[i] = (parent_1[i]+parent_2[i])/2
    else:
      child[i] = parent_2[i]
  
  #Regresamos el individuo generado
  return child

## Mutación modificada

In [197]:
def mutation_modified(i, n, epsilon, tau):
  #Mutamos el tamaño de paso
  mutation_sigma = i[-1]*(e**(tau*np.random.normal(0, 1, 1)))
  #Verificamos que el nuevo valor no sea menor a épsilon
  if mutation_sigma[0] < epsilon:
    mutation_sigma[0] = epsilon

  #Mutamos las variables de decisión
  mutation_x = i[:n] + (mutation_sigma[0]*np.random.normal(0, 1, n))

  # En lugar de hacer una asignación fija, vamos a implementar una asignación variable
  aux_less = -1*int(30/2)
  aux_plus = int(30/2)
  for i in range(len(mutation_x)): 
    if mutation_x[i] < -30: 
      mutation_x[i] = aux_less
      if (aux_less > -30) == False: 
        aux_less = -20
      else: 
        aux_less += -2
    
    if mutation_x[i] > 30: 
      mutation_x[i] = aux_plus
      if (aux_plus > 30) == True:
        aux_plus = 20
      else: 
        aux_plus += 2

  return np.concatenate((mutation_x, mutation_sigma))

## Creación de hijos

In [199]:
def createChild_plus(parent_1, parent_2, n, epsilon, tau):
  new_ind = localDiscreteCrossover_modified(parent_1, parent_2)
  new_ind = mutation_modified(new_ind, n, epsilon, tau)

  #Regresamos el nuevo individuo, junto con su valor de apitud
  return [new_ind, fitness(new_ind[:n])]

## ($\mu, \lambda$)-EE Modificada

In [200]:
def EvolutionStrategies_with_plus(n, G, epsilon=0.01, tau=None, mu_=100, lambda_=700):
  #Si el usuario no define tau,
  #se usa el valor recomendado en la literatura
  if tau == None:
    tau = 1/sqrt(n)

  population = getInitialPopulation_plus(mu_, n)
  #Generamos los índices de los individuos para poder elegirlos
  #aleatoriamente
  population_idx = range(mu_)

  #Repetimos el proceso por G generaciones
  for _ in range(G):
    offspring = []
    #Generamos lambda_ hijos
    for i in range(lambda_):
      #Seleccionamos a los padres
      parents_idx = choices(population_idx, k=2)
      parent_1 = population[population_idx[0]][0]
      parent_2 = population[population_idx[1]][0]
      #Creamos el hijo y lo añadimos a la población de hijos
      offspring.append( createChild_plus(parent_1, parent_2, n, epsilon, tau) )
    
    #Ordenamos del mejor individuo al peor individuo
    offspring.sort(key=lambda x: x[-1], reverse=True)
    #Seleccionamos a los mu_ mejores
    population = offspring[:mu_].copy()

  
  #Regresamos el mejor individuo
  return population[0][0], -population[0][1]

# Ejercicio 3
---
Permitir al usuario realizar M ejecuciones del algoritmo de EE, implementado en el punto anterior para resolver la instancia del problema de Ackley. 

Reportar: Mejor solución, Peor solución, La media de soluciones, desviación estandar del valor de la función objetivo

In [208]:
from dotenv import dotenv_values
if __name__=="__main__":
  variables = dotenv_values("11-EE.txt")
  N = int(variables["plusN"])
  MU_ = int(variables["plusMu"]) 
  LAMBDA_ = int(variables["plusLambda"]) 
  GG = int(variables["plusG"]) 
  TAU = 1/sqrt(n) 
  EPSILON = float(variables["plusEpsilon"]) 

  M = int(input("Número de ejecuciones: "))

  print(N)
  solutions = {}
  conteo = 0
  while conteo != M:
    current_xs, current_fx = EvolutionStrategies_with_plus(n=N, G=GG, epsilon=EPSILON,tau=TAU, mu_=MU_, lambda_=LAMBDA_)
    xs = current_xs[:N]
    fx = current_fx
    solutions[tuple(xs)] = fx
    conteo += 1  

  
  # TODO: Ordenamos de menor a mayor por fitness
  solutions = sorted(solutions.items(), key=lambda x: x[1])
  #print(solutions)
  #print(solutions[0]) #best
  #print(solutions[-1]) #faild
  
  # MEJOR SOLUCION
  print(f"\nMEJOR SOLUCION\nXs: {solutions[0][0][:-1]}\nfx: {solutions[0][1]}")
  # PEOR SOLUCION 
  print(f"\nPEOR SOLUCION\nXs: {solutions[-1][0][:-1]}\nfx: {solutions[-1][1]}")
  # MEDIANAS
  print("\nMEDIANA: Xs&fx")
  tmp_X = []
  tmp_F = []
  for i in range(len(solutions)): 
    tmpX = solutions[i][0][:-1]
    tmpF = solutions[i][1]
    tmp_X.append(tmpX)
    tmp_F.append(tmpF)


  num_ele = len(tmp_X[0])
  aux = [0 for i in range(num_ele)]
  
  for element in tmp_X:
    for i in range(num_ele):
      aux[i] += element[i]

  for i in range(len(aux)):
    aux[i] = aux[i]/len(tmp_X)

  # Media de ambos
  print(f"Xs&fx: {aux} & {sum(tmp_F)/len(tmp_F)}")
  
  # Media del valor de la f objetivo
  print("\nMEDIA FUNCION OBJETIVO")
  print(f"Media: {sum(tmp_F)/len(tmp_F)}")

  # Desviación estánda de la f objetivo
  print("\nDESVIACION ESTANDAR FUNCION OBJETIVO")
  print(f"SD: {np.std(tmp_F)}")  

Número de ejecuciones: 10
10

MEJOR SOLUCION
Xs: (0.004420046587569724, 0.00035445490769713895, 0.004402534446553326, 0.00014611252983356754, 0.000172854842535602, 0.009088904085331414, -0.0018491751790590946, -0.002416399144102829, 0.006050284694682531)
fx: 0.017466030363323615

PEOR SOLUCION
Xs: (-19.998994302737838, -3.9970848747708967, -0.0023164593762708564, -1.0028266951330405, 5.001517168997477, -17.996775554301692, 6.988993889518492, 19.99069474715385, -0.9978270129615832)
fx: 17.890848805206414

MEDIANA: Xs&fx
Xs&fx: [-1.6994488066475821, -1.6984491923289482, -0.398483036213164, 1.4979317807581385, -0.1997073087455182, -0.29671741469677926, 1.3980132590082726, 1.796938309170787, -1.399356606663876] & 3.525860155614563

MEDIA FUNCION OBJETIVO
Media: 3.525860155614563

DESVIACION ESTANDAR FUNCION OBJETIVO
SD: 7.014478993012486


# Ejercicio 4
---
Me ha resultado muy dificil hacer una tabla comparativa aquí con markdown, lo siento mucho si esto te complica más el revisar los resultados. No es mi intención dificultarte la evaluación. 



## Resultados

# ALGORITMO (1+1)-EE

## n = 2

MEJOR SOLUCION
Xs: (-1.142502665578857,)
fx: 2.4200543113066932

MEDIANA: Xs&fx
Xs&fx: [-0.3362261291596411] & 19.89657893084126

## n = 5

MEJOR SOLUCION
Xs: (2.251628513294725, -0.8552266373665476, 10.509473886693849, -10.336022599432713)
fx: 14.06043338102895

MEDIANA: Xs&fx
Xs&fx: [0.12741528041417274, -0.34705642397619574, 0.15826665944670285, -0.42751948945306395] & 20.761250621021976

## n = 7

MEJOR SOLUCION
Xs: (-0.3942074350828388, 5.977341767189939, 0.8045857815256243, 6.301963576604272, 4.250286311734453, -1.6181860251572715)
fx: 13.893407370011952

MEDIANA: Xs&fx
Xs&fx: [0.3825494665775337, -0.12847038195474642, 0.11402759456928693, -0.2310110649755229, -0.07415102120264548, -0.32259527614265887] & 20.878749797387382

## n = 10

MEJOR SOLUCION
Xs: (-2.044246570159379, 17.88231351065758, -6.052851320747088, -1.1693274828589078, -0.9154921722675008, 7.962487770169621, -2.9105446028522426, -6.105693861656952, 1.1571605460194185)
fx: 17.547398709030844

MEDIANA: Xs&fx
Xs&fx: [0.010372363820764036, -0.02250920550874845, -0.18706328010297152, 0.4985264746541095, 0.08681879516998701, -0.27355355064089315, 0.08632618368508485, -0.36110568281307387, 0.031095853796096355] & 20.946555311456187

# ALGORITMO (mu, lambda)-EE
---
Para este algoritmo tuve que reducir las ejecución de 5000 a 100 debido a que
para n = 2 después de 15 minutos aún no terminaba. La he bajado a 10 ejecuciones ya
que para n = 2 después de seis minutos no terminaba y no queria perder mucho tiempo
entre un caso y otro

## n = 2

MEJOR SOLUCION
Xs: (-2.850511722755683e-06,)
fx: 0.00015762266398811065

MEDIANA: Xs&fx
Xs&fx: [-7.00993642101619e-05] & 0.0017123825916592673

## n = 5

MEJOR SOLUCION
Xs: (-0.0007418717877417545, -0.0010435098095047547, -0.0017318702315364293, -0.0008017086674698879)
fx: 0.005946192550876805

MEDIANA: Xs&fx
Xs&fx: [-0.0011901986994583165, 0.0004330298117639753, -0.0011428405068369804, -0.0002689190084993846] & 0.009109659856281826


## n = 7

MEJOR SOLUCION
Xs: (0.00013798782180813644, -0.002419827427409884, 0.0014021533860532648, 0.001052838240213387, 9.420560686191522e-05, -0.0004660784956039138)
fx: 0.008765996129266984

MEDIANA: Xs&fx
Xs&fx: [0.00033167664554998466, 0.0007614059208538299, 0.00020789993593843268, 0.0002276489383165554, -0.0011807471092961327, -0.0007901537648412232] & 0.012093297799161195


## n = 10

MEJOR SOLUCION
Xs: (-0.007856224531891857, -0.004900643004742735, 0.0023537358139212325, -0.0008414020724393399, 0.002150847399051097, 0.0046581898315798944, -0.0012226454308359954, 0.001062616874116001, 0.002619402286077113)
fx: 0.014992679304719037

MEDIANA: Xs&fx
Xs&fx: [-0.0016897703135800156, -0.0015113848830057494, -0.001775757451882999, 0.0014822654723971043, 0.002669917249006662, 0.001753149609581519, 0.00021638853540399913, -0.0006085943719727511, 0.0001282185673733577] & 0.018763338761734704

# ALGORITMO (mu, lambda)-EE con modificaciones implementadas

## n = 2

MEJOR SOLUCION
Xs: (-0.000207548674756605,)
fx: 0.0007028745808530523

MEDIANA: Xs&fx
Xs&fx: [-0.19975958885452802] & 1.5898436637876647

## n = 5

MEJOR SOLUCION
Xs: (-0.0009790976897849375, -0.0014898144287687383, 0.0004773494041373778, 0.001506930863386827)
fx: 0.005249789878320588

MEDIANA: Xs&fx
Xs&fx: [0.2998739155678475, -3.39894248470828, 4.6986586905324375, 0.5985083868005623] & 5.572852708420219

## n = 7

MEJOR SOLUCION
Xs: (-0.004379147750717403, 0.0019322258693806448, -0.004068392658695756, 0.000688083632791084, 0.0002639492455335274, -0.0017949343249720532)
fx: 0.01076345269676482

MEDIANA: Xs&fx
Xs&fx: [-0.299228788074312, -3.1961626549137905, 3.5974338935273225, -3.09804691923341, -2.399372987480376, 4.296222646805348] & 6.293547777969152

## n = 10

MEJOR SOLUCION
Xs: (0.004420046587569724, 0.00035445490769713895, 0.004402534446553326, 0.00014611252983356754, 0.000172854842535602, 0.009088904085331414, -0.0018491751790590946, -0.002416399144102829, 0.006050284694682531)
fx: 0.017466030363323615

MEDIANA: Xs&fx
Xs&fx: [-1.6994488066475821, -1.6984491923289482, -0.398483036213164, 1.4979317807581385, -0.1997073087455182, -0.29671741469677926, 1.3980132590082726, 1.796938309170787, -1.399356606663876] & 3.525860155614563


Según parece, todos los algoritmos convergen mucho más rápido a menor número de variables pero el número de ejecuciones juega un papel importante también.